include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,1049 @@
+    can be used to validate the underlying string
+    before constructing a `decoded_view`.
+
+*/
+class decoded_view
+{
+    char const* p_{nullptr};
+    std::size_t n_{0};
+    std::size_t dn_{0};
+    bool plus_to_space_{true};
+
+    using traits_type = std::char_traits<char>;
+
+public:
+    /** Type of a decoded character
+    */
+    using value_type = char;
+
+    /// @copydoc value_type
+    using reference = char;
+
+    /// @copydoc value_type
+    using const_reference = char;
+
+    /** The unsigned integer type used to represent size.
+    */
+    using size_type = std::size_t;
+
+    /** The signed integer type used to represent differences.
+    */
+    using difference_type = std::ptrdiff_t;
+
+    /** A read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+     */
+#ifdef BOOST_URL_DOCS
+    using const_iterator = __see_below__;
+#else
+    class const_iterator;
+#endif
+
+    /// @copydoc const_iterator
+    using iterator = const_iterator;
+
+    /** A reverse read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+     */
+    using reverse_iterator = std::reverse_iterator<iterator>;
+
+    /// @copydoc reverse_iterator
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    /** Constructor
+
+        Default-constructed objects represent
+        the empty string.
+
+     */
+    BOOST_CONSTEXPR
+    decoded_view(
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(nullptr)
+        , n_(0)
+        , dn_(0)
+        , plus_to_space_(opt.plus_to_space)
+    {}
+
+    /** Constructor
+     */
+    decoded_view(
+        char const* str,
+        pct_decode_opts const& opt = {}) noexcept
+        : decoded_view(string_view(str), opt)
+    {}
+
+    /** Constructor
+     */
+    decoded_view(
+        char const* str,
+        size_type len,
+        pct_decode_opts const& opt = {}) noexcept
+        : decoded_view(string_view(str, len), opt)
+    {}
+
+    /** Constructor
+     */
+#ifndef BOOST_URL_DOCS
+    template <
+        class End,
+        typename std::enable_if<std::is_same<End, char const*>::value, int>::type = 0>
+    BOOST_CXX14_CONSTEXPR
+    decoded_view(
+        char const* first,
+        End last,
+        pct_decode_opts const& opt = {}) noexcept
+#else
+    BOOST_CXX14_CONSTEXPR
+    decoded_view(
+        char const* first,
+        char const* last,
+        pct_decode_opts const& opt = {}) noexcept
+#endif
+        : decoded_view(string_view(first, last), opt)
+    {
+        BOOST_ASSERT(last - first >= 0);
+    }
+
+    /** Constructor
+     */
+    decoded_view(
+        urls::string_view str,
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(str.data())
+        , n_(str.size())
+        , dn_(pct_decode_bytes_unchecked(str))
```

Comment:
This is a case where you don't want to initialize the member in the ctor-init list. For two reasons: 1. you are passing untrusted input to a function that does not perform validation, and 2. you are calculating the decoded size twice (`validate_pct_encoding` returns the size of the resulting decoded string).